aws/credential_cache.go

```diff
@@ -41,8 +98,17 @@ func (p *CredentialsCache) singleRetrieve() (interface{}, error) {
 		return *creds, nil
 	}
 
-	creds, err := p.Provider.Retrieve(context.TODO())
+	creds, err := p.options.Provider.Retrieve(context.TODO())
 	if err == nil {
+		if creds.CanExpire {
+			randFloat64, err := sdkrand.CryptoRandFloat64()
+			if err != nil {
+				return Credentials{}, err
+			}
+			jitter := time.Duration(randFloat64 * p.options.ExpiryWindowJitterFrac * float64(p.options.ExpiryWindow))
+			creds.Expires = creds.Expires.Add(-(p.options.ExpiryWindow - jitter))
+		}
+
 		p.creds.Store(&creds)
```

Comment:
Is there anyway for this to log, or report, when the a code bug sets ExpiryWindow is so large a value that it causes the credentials to immediately expire? Incorrect expiry windows being set seem to be a good way for apps to get stuck in a refresh, or credential expire loop without a clear indication why.

Reply:
There are two issues I would foresee with this:
1. this condition can't be verified up front during construction of the provider cache in a reliable fashion. This is due to the credential expiration time not being known until the credentials are retrieved. Even so, subsequent refreshes of the credential can still cause the situation. So even if you were to pre-warm the cache on construction and verify this condition does not occur, it would not hold true for subsequent retrievals.
2. Logging would require the cache provider to be configured with the logger, and subsequently could make the situation worse in that you would have a log storm of messages about this condition despite a still functioning application (a bit with unnecessary latency due to need to refresh credentials before a call).